### PR TITLE
ICU-22093 Polish for SimpleNumberFormatter

### PR DESCRIPTION
**Pull request by sffc**
_No hurry to review this; I don't expect a review until January_

CC @Manishearth 

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22093
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

[ICU-22093]: https://unicode-org.atlassian.net/browse/ICU-22093?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ICU-1234]: https://unicode-org.atlassian.net/browse/ICU-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ICU-1234]: https://unicode-org.atlassian.net/browse/ICU-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: icu4c/source/i18n/unicode/usimplenumberformatter.h [hunk 1]
```diff
@@ -175,14 +208,13 @@ usnumf_openForLocaleAndGroupingStrategy(
 /**
  * Formats a number using this SimpleNumberFormatter.
  *
- * The USimpleNumber is adopted and must not be freed after calling this function,
- * even if the function sets an error code. If you use LocalUSimpleNumberPointer,
- * call `.orphan()` when passing it to this function.
+ * The USimpleNumber is cleared after calling this function. It can be re-used via
+ * usnum_setToInt64.
```
Comment (richgillam) [inline icu4c/source/i18n/unicode/usimplenumberformatter.h:212]: What happens if the caller doesn't call `unum_setToInt64()`? Do they get an error?  Do they get the previous value formatted again?

Comment (richgillam) [reply icu4c/source/i18n/unicode/usimplenumberformatter.h:212]: Never mind; I see from the unit tests that you just return an `ILLEGAL_ARGUMENT_ERROR`, and that seems fine to me.

**Review by richgillam — APPROVED** [review]
I like your improvements here.

File: icu4c/source/test/cintltst/unumberformattertst.c [hunk 2]
```diff
@@ -212,6 +216,26 @@ static void TestExampleCode() {
 }
 
 
+static void TestSimpleNumberFormatter() {
```
Comment (markusicu) [inline icu4c/source/test/cintltst/unumberformattertst.c:219]: This is good, but please check that you call each new C function at least once from cintltst.

**Review by markusicu — COMMENTED** [review]
lgtm2 but
- let's confirm the API changes in the team meeting
- @sffc please confirm that all new C functions are called at least once from cintltst

**Comment by markusicu** [comment]
> * let's confirm the API changes in the team meeting

done, approved

> * @sffc please confirm that all new C functions are called at least once from cintltst

please do check

File: icu4c/source/test/cintltst/unumberformattertst.c (hunk 2 above)
Comment (sffc) [reply icu4c/source/test/cintltst/unumberformattertst.c:219]: Okay. I was calling every function, but not from the C test. I am now calling every function from the C test.

File: icu4c/source/test/intltest/numbertest_simple.cpp [hunk 3]
```diff
@@ -185,21 +185,37 @@ void SimpleNumberFormatterTest::testCAPI() {
...
     assertEquals("",
         u"44",
         ufmtval_getString(unumf_resultAsValue(uresult.getAlias(), status), nullptr, status));
 
-    unumber.adoptInstead(usnum_openForInt64(2335, status));
+    // Can't re-use a number without setting it again
+    usnumf_format(uformatter.getAlias(), unumber.getAlias(), uresult.getAlias(), status);
+    status.expectErrorAndReset(U_ILLEGAL_ARGUMENT_ERROR);
+    usnum_multiplyByPowerOfTen(unumber.getAlias(), 0, status);
+    status.expectErrorAndReset(U_INVALID_STATE_ERROR);
```
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4c/source/test/intltest/numbertest_simple.cpp]: Hmm. I think it actually makes sense the way you have it.
- Manipulating a bogus object, "invalid state" seems to make sense.
- Passing a bogus object into a formatter, "illegal argument" sounds about right.

I haven't checked whether we are consistent in how we use "invalid state"...

File: icu4c/source/test/cintltst/unumberformattertst.c (hunk 2 above)
Comment (markusicu) [reply icu4c/source/test/cintltst/unumberformattertst.c:219]: Thanks!

File: icu4c/source/test/cintltst/unumberformattertst.c [hunk 4]
```diff
@@ -212,6 +216,64 @@ static void TestExampleCode() {
...
+    // This is the example in usimplenumberformatter.h
```
Comment (markusicu) [inline icu4c/source/test/cintltst/unumberformattertst.c:220]: optional: Consider either adding another comment for where this .h example ends, or else moving the example code into its own block or function.

File: icu4c/source/test/cintltst/unumberformattertst.c [hunk 5]
```diff
@@ -212,6 +216,64 @@ static void TestExampleCode() {
...
+ * For more control over the formatting, use USimpleNumber.
+ *
+ * @draft ICU 73
+ */
+U_CAPI void U_EXPORT2
+usnumf_formatInt64(
+    const USimpleNumberFormatter* uformatter,
+    int64_t value,
+    UFormattedNumber* uresult,
+    UErrorCode* ec);
```
Comment (markusicu) [inline icu4c/source/test/cintltst/unumberformattertst.c:274]: why is this here? redundant with the public header?

File: icu4c/source/test/intltest/numbertest_simple.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_simple.cpp:197]: OK, I changed it to U_ILLEGAL_ARGUMENT_ERROR and now changed it back to U_INVALID_STATE_ERROR

File: icu4c/source/test/cintltst/unumberformattertst.c (hunk 4 above)
Comment (sffc) [reply icu4c/source/test/cintltst/unumberformattertst.c:223]: Split it into two test functions

File: icu4c/source/test/cintltst/unumberformattertst.c (hunk 5 above)
Comment (sffc) [reply icu4c/source/test/cintltst/unumberformattertst.c]: Oops, removed

File: icu4c/source/test/intltest/numbertest_simple.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4c/source/test/intltest/numbertest_simple.cpp:197]: Similar use of U_INVALID_STATE_ERROR in ubidiimp.h, alphaindex.cpp, selfmt.cpp, and some other places.

**Review by markusicu — COMMENTED** [review]
lgtm tnx!
Please squash, and fix the commit message.

**Review by markusicu — APPROVED** [review]
lgtm tnx
